testing/trino-product-tests/src/main/java/io/trino/tests/hive/TestHiveStorageFormats.java

```diff
@@ -536,6 +546,35 @@ public void testStructTimestamps(StorageFormat format)
         onHive().executeQuery(format("DROP TABLE %s", tableName));
     }
 
+    @Test(dataProvider = "storageFormatsWithNanosecondPrecisionExceptParquet", groups = STORAGE_FORMATS)
+    public void testStructTimestampsFromTrino(StorageFormat format)
+    {
+        String tableName = createStructTimestampTable("trino_struct_timestamp", format);
+        setAdminRole(onPresto().getConnection());
+
+        for (HiveTimestampPrecision precision : HiveTimestampPrecision.values()) {
+            setTimestampPrecision(precision);
+
+            onPresto().executeQuery(format(
+                    "INSERT INTO %s VALUES (%s)",
+                    tableName,
+                    TIMESTAMPS_FROM_TRINO.stream()
+                            .filter(t -> t.getPrecision() == precision)
```

Comment:
Why filter here? what would happen if not filtered?

Reply:
This inserts each timestamp at the precision it was specified for. Otherwise, they wouldn't round as expected.

To make this clearer, I've changed it to use explicit grouping logic.